Sources/Web3Core/KeystoreManager/BIP44.swift

```diff
@@ -0,0 +1,126 @@
+    derived directly using the derive function of ``HDNode``. This function needs to query the blockchain history when throwOnWarning is `true`, so it can throw
+    network errors.
+    - Parameter path: valid BIP44 path.
+    - Parameter throwOnWarning: `true` to use
+    [Account Discovery](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery) standard,
+    otherwise it will dervive the key using the derive function of ``HDNode``.
+    - Throws: ``BIP44Error.warning`` if the child key shouldn't be used according to
+    [Account Discovery](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery) standard.
+    - Returns: an ``HDNode`` child key for the provided `path` if it can be created, otherwise `nil`
+    */
+    func derive(path: String, throwOnWarning: Bool, transactionChecker: TransactionChecker) async throws -> HDNode?
+}
+
+public enum BIP44Error: LocalizedError, Equatable {
+    /// The selected path doesn't fulfill BIP44 standard, you can derive the root key anyway
+    case warning
+
+    public var errorDescription: String? {
+        switch self {
+        case .warning:
+            return "Couldn't derive key as it doesn't have a previous account with at least one transaction"
+        }
+    }
+}
+
+public protocol TransactionChecker {
+    /**
+    It verifies if the provided address has at least one transaction
+    - Parameter address: to be queried
+    - Throws: any error related to query the blockchain provider
+    - Returns: `true` if the address has at least one transaction, `false` otherwise
+    */
+    func hasTransactions(address: String) async throws -> Bool
+}
+
+extension HDNode: BIP44 {
+    public func derive(path: String, throwOnWarning: Bool = true, transactionChecker: TransactionChecker) async throws -> HDNode? {
+        guard throwOnWarning else {
+            return derive(path: path, derivePrivateKey: true)
+        }
+        guard let account = path.accountFromPath else {
+            return nil
+        }
+        if account == 0 {
+            return derive(path: path, derivePrivateKey: true)
+        } else {
+            for searchAccount in 0..<account {
+                let maxUnusedAddressIndexes = 20
+                var hasTransactions = false
+                for searchAddressIndex in 0..<maxUnusedAddressIndexes {
+                    if let searchPath = path.newPath(account: searchAccount, addressIndex: searchAddressIndex),
+                    let childNode = derive(path: searchPath, derivePrivateKey: true),
+                    let ethAddress = Utilities.publicToAddress(childNode.publicKey) {
+                        hasTransactions = try await transactionChecker.hasTransactions(address: ethAddress.address)
+                        if hasTransactions {
+                            break
```

Comment:
I suppose the intention of this `break` was to reach go right to `return derive(path: path, derivePrivateKey: true)` on line 75? Please, correct me if I'm wrong.

If so, then it is not what happens. You only exit the inner for-loop and continue whit the next round of the outer for-loop.

Consider this example for simplicity:

<img width="1508" alt="Screenshot 2023-01-18 at 22 07 41" src="https://user-images.githubusercontent.com/36865532/213283742-84c5b412-dd26-4184-91ed-14757e02d6ac.png">

To fix that issue (and I think it is an issue if I understand BIP44 correctly) you can add a label to the outer for-loop and get the following result:

<img width="994" alt="Screenshot 2023-01-18 at 22 09 01" src="https://user-images.githubusercontent.com/36865532/213283979-33eecc23-104d-4243-8b96-a4c517b38a8d.png">

Reply:
See the difference on the first and the second screenshot of the output of `print` function call.

Reply:
The intention is to break only the internal for loop, I will try to explain the reason:
BIP44 defines that we should check at least 20 addressIndex before saying that an account index doesn't have transactions, so the idea here is to loop until 20 unless we find an addressIndex with balance, in this case we can break and continue with the outer loop to analize the next account index

Reply:
I'll have to re-read BIP44 standard as I'm still unsure about a few things.

You said "unless we find an addressIndex with balance" but shouldn't it be "... with transaction"? I don't see you checking balances here.

Soon I'll get back to you with either more questions (hopefully not) or an approved PR.

Reply:
Yep, it's "transactions", not "balances": 

> Please note that the algorithm works with the transaction history, not account balances, so you can have an account with 0 total coins and the algorithm will still continue with discovery.

Typos happen, just clarifying for everyone who'll read this thread.

Reply:
Approved.

Reply:
You are right about balance, apologies, the word is transactions